src/python/pants/engine/README.md

```diff
@@ -17,63 +17,69 @@ Once the engine is instantiated with a valid set of `@rule`s, a caller can synch
-Product given a single `Int` input.
+Product given a single `Int` argument.
 
 ```python
-@rule(StringType, [IntType])
+@rule(str, [int])
 def int_to_str(an_int):
-  return '{}'.format(an_int)
+  return str(an_int)
 ```
 
-The first argument to the `@rule` decorator is the Product (ie, return) type for the `@rule`. The
-second argument is a list of parameter selectors that declare the types of the input parameters for
-the `@rule`. In this case, because the Product type is `StringType` and there is one parameter
-selector (for `IntType`), this `@rule` represents a conversion from `IntType` to `StrType`, with no
-other inputs.
+The first argument to the `@rule` decorator is the `Product` (ie, return) type for the `@rule`. The
+second argument is a list of "parameter selectors" that declare the types of the input parameters for
+the `@rule`. In this case, because the `Product` type is `str` and there is one parameter
+selector (for `int`), this `@rule` represents a conversion from `int` to `str`, with no other inputs.
 
-When the engine statically checks whether it can use this `@rule` to create a string for a
-Subject, it will first see whether there are any ways to get an IntType for that Subject. If
-the subject is already of `type(subject) == IntType`, then the `@rule` will be satisfiable without
-any other dependencies. On the other hand, if the type _doesn't_ match, the engine doesn't give up:
-it will next look for any other registered `@rule`s that can compute an IntType Product for the
-Subject (and so on, recursively).
+When the engine encounters this `@rule` while compiling the rule graph for `str`-producing-`@rules`,
+it will next go hunting for the dependency `@rule` that can produce an `int` using the fewest number
+of `Params`. For example, if there was an `@rule` that could produce an `int` without consuming any
+`Params` at all (ie, a singleton), then that `@rule` would always be chosen first. If all `@rules` to
+produce `int`s required at least one `Param`, then the engine would next see whether the input `Params`
+contained an `int`, or whether there were any `@rules` that required only one `Param`... then two
+`Params`, and etc.
+
+In cases where this search detects any ambiguity (generally because there are two or more `@rules` that
+can provide the same product with the same number of parameters), rule graph compilation will fail with
+a useful error message.
 
 ### Datatypes
 
-In practical use, using basic types like `StringType` or `IntType` does not provide enough
-information to disambiguate between various types of data. So declaring small `datatype`
-definitions to provide a unique and descriptive type is strongly recommended:
+In practical use, builtin types like `str` or `int` do not provide enough information to disambiguate
+between various types of data, so declaring small `datatype` definitions to provide a unique and
+descriptive type is highly recommended:
```

Comment:
Why do we need to do this in the first place? Worth adding a phrase like `to add as the Product and Params for your @rule definition`.